### PR TITLE
Method validateIds in AbstractRequestBase

### PR DESCRIPTION
**Pull request by xrosecky**
We need some refactoring to disable validation of IDs against session in our code and this PR makes it possible.

### CODE REVIEW

**Comment by demiankatz** [comment]
This seems reasonable to me as a non-backward-breaking improvement. I'm a little surprised that there's not parallel logic to this for the ILL and storage retrieval request modules, but I couldn't find any. Is that cause for concern, or are holds just more complicated than those other types of requests?

**Comment by xrosecky** [comment]
Yes, I was also surprised that the code from AbstractRequestBase is used only in HoldsController.

**Comment by demiankatz** [comment]
Any thoughts, @EreMaijala? (I can dig into this in more detail after the Summit if research is needed).

**Comment by EreMaijala** [comment]
Validation is probably one of things that should have been added to ILL and storage retrieval request modules, and that's why the methods are in the base class. I probably just forgot to do that..

**Comment by demiankatz** [comment]
There is also outstanding work to refactor more logic out of the MyResearchController. Maybe we can find time to do these things as part of the same project. Perhaps at least worth a JIRA ticket, and maybe some discussion during roadmapping tomorrow. (Getting rid of MyResearch is still a "nice-to-have" on my list).

**Review by EreMaijala — APPROVED** [review]
If we do these changes in dev-9.0, I'm fine with this. However, I'm not sure moving `getValidIds` from public to protected in dev is a good idea.

**Comment by EreMaijala** [comment]
@demiankatz Now that I think of it, the idea when I was working on ILL and SR requests could have been that the driver would bear more responsibility on doing the right thing, but validating IDs still wouldn't hurt. I created https://vufind.org/jira/browse/VUFIND-1521 as a placeholder. It's true that refactoring of MyResearch would be nice, but I don't think that's necessary for adding the validation.

File: module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php [hunk 1]
```diff
@@ -136,13 +136,16 @@ public function rememberValidId($id)
...
-     * @return array
+     * @param array $ids IDs to validate
+     *
+     * @return bool
      */
-    public function getValidIds(): array
+    public function validateIds($ids): bool
     {
-        return $this->getSession()->validIds ?? [];
+        return array_diff($ids, $this->getValidIds());
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php]: Should we typecast this to (bool) or use some other check that creates a bool in order to be a little more strict about our typing?

File: module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php [hunk 2]
```diff
@@ -354,4 +357,14 @@ protected function getDateFromArray($dateArray)
             date('Y') + $y
         );
     }
+
+    /**
+     * Get remembered valid IDs
+     *
+     * @return array
+     */
+    protected function getValidIds(): array
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php:366]: To @EreMaijala's point, if we keep this public here, we avoid backward compatibility breaks. Would you like to change it back to public so we can merge to dev, and then I can make it protected in dev-9.0, or do you prefer to simply merge this change directly to dev-9.0?

**Review by demiankatz — COMMENTED** [review]
Thanks for the input, @EreMaijala. In light of that, I took another look, which also led me to notice one other very minor thing. See comments below.

Also, regarding the MyResearch refactor, I agree that it's certainly not a prerequisite for adding more validation... just thought that one project might help to justify the completion of the other in the same chunk of work. :-)

File: module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php (hunk 2 above)
Comment (xrosecky) [reply module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php:366]: @demiankatz You can merge this change only to dev-9.0.

**Comment by demiankatz** [comment]
Thanks, @xrosecky; I have changed the base and will merge this to dev-9.0 now.

**Pull request merged**